File: lib/src/main/java/com/sd/lib/http/RequestTask.java
package com.sd.lib.http;

import com.sd.lib.http.callback.IUploadProgressCallback;
import com.sd.lib.http.callback.RequestCallback;
import com.sd.lib.http.utils.HttpLog;
import com.sd.lib.http.utils.TransmitParam;

final class RequestTask extends FTask implements IUploadProgressCallback
{
    private final IRequest mRequest;
    private final RequestCallback mRequestCallback;

    private final Callback mCallback;

    private volatile boolean mIsCancelled = false;
    private boolean mIsStartNotified = false;

    public RequestTask(IRequest request, RequestCallback requestCallback, Callback callback)
    {
        mRequest = request;
        mRequestCallback = requestCallback;
        mCallback = callback;

        mRequest.setUploadProgressCallback(this);
    }

    private String getLogPrefix()
    {
        return String.valueOf(this);
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning)
    {
        HttpLog.e(getLogPrefix() + " cancel called start");

        final boolean result = super.cancel(mayInterruptIfRunning);
        if (result)
        {
            // 正常情况下，父类的cancel执行之后onCancel中已经赋值，这里再次赋值
            mIsCancelled = true;
        }

        HttpLog.e(getLogPrefix() + " cancel called result:" + result);
        return result;
    }

    @Override
    protected void onRun() throws Exception
    {
        HttpLog.i(getLogPrefix() + " 1 onRun");
        if (checkCancel())
            return;

        runOnUiThread(mStartRunnable);
        synchronized (RequestTask.this)
        {
            if (!mIsStartNotified)
            {
                // 等待开始回调完成
                HttpLog.i(getLogPrefix() + " wait start...");
                RequestTask.this.wait();
                HttpLog.i(getLogPrefix() + " wait finish");
            }
        }

        HttpLog.i(getLogPrefix() + " 2 before execute");
        if (checkCancel())
            return;

        final IResponse response = mRequest.execute();

        HttpLog.i(getLogPrefix() + " 3 after execute");
        if (checkCancel())
            return;

        mRequestCallback.setResponse(response);
        mRequestCallback.onSuccessBackground();

        HttpLog.i(getLogPrefix() + " 4 after onSuccessBackground");
        if (checkCancel())
            return;

        HttpLog.i(getLogPrefix() + " 5 success");
        runOnUiThread(mSuccessRunnable);
    }

    private boolean checkCancel()
    {
        if (mIsCancelled)
        {
            HttpLog.i(getLogPrefix() + " check cancelled !!!");
            return true;
        }
        return false;
    }

    private final Runnable mStartRunnable = new Runnable()
    {
        @Override
        public void run()
        {
            if (mIsCancelled)
            {
                // 如果被取消的话，此处不通知开始事件
                HttpLog.e(getLogPrefix() + " start runnable but is cancelled");
                return;
            }

            synchronized (RequestTask.this)
            {
                if (!mIsStartNotified)
                {
                    mIsStartNotified = true;
                    HttpLog.i(getLogPrefix() + " notify onStart");
                    mRequestCallback.onStart();
                }

                RequestTask.this.notifyAll();
            }
        }
    };

    private final Runnable mSuccessRunnable = new Runnable()
    {
        @Override
        public void run()
        {
            if (mIsCancelled)
            {
                HttpLog.e(getLogPrefix() + " success runnable but is cancelled");
                return;
            }

            mRequestCallback.onSuccessBefore();
            mRequestCallback.onSuccess();
        }
    };

    @Override
    protected void onError(final Exception e)
    {
        HttpLog.i(getLogPrefix() + " onError:" + e);
        runOnUiThread(new Runnable()
        {
            @Override
            public void run()
            {
                mRequestCallback.onError(e);
            }
        });
    }

    @Override
    protected void onCancel()
    {
        super.onCancel();

        mIsCancelled = true;
        HttpLog.i(getLogPrefix() + " onCancel");

        runOnUiThread(new Runnable()
        {
            @Override
            public void run()
            {
                mRequestCallback.onCancel();
            }
        });
    }

    @Override
    protected void onFinish()
    {
        super.onFinish();
        HttpLog.i(getLogPrefix() + " onFinish");
        runOnUiThread(new Runnable()
        {
            @Override
            public void run()
            {
                mRequestCallback.onFinish();
            }
        });

        mCallback.onFinish(this);
    }

    @Override
    public void onProgressUpload(TransmitParam param)
    {
        mRequestCallback.onProgressUpload(param);
    }

    public interface Callback
    {
        void onFinish(RequestTask task);
    }
}

File: lib/src/main/java/com/sd/lib/http/RequestTask.kt
package com.sd.lib.http

import com.sd.lib.http.callback.IUploadProgressCallback
import com.sd.lib.http.callback.RequestCallback
import com.sd.lib.http.exception.HttpException
import com.sd.lib.http.utils.HttpLog
import com.sd.lib.http.utils.HttpUtils
import com.sd.lib.http.utils.TransmitParam
import kotlinx.coroutines.*

internal abstract class RequestTask {
    private val _request: IRequest
    private val _requestCallback: RequestCallback

    /** 任务 */
    private var _job: Job? = null

    @Volatile
    private var _isStartNotified: Boolean = false

    @Volatile
    private var _isResultNotified: Boolean = false

    constructor(request: IRequest, requestCallback: RequestCallback) {
        _request = request
        _requestCallback = requestCallback

        request.uploadCallback = object : IUploadProgressCallback {
            override fun onProgressUpload(params: TransmitParam) {
                _requestCallback.notifyProgressUpload(params)
            }
        }
    }

    /** 日志前缀 */
    private val _logPrefix: String by lazy { this@RequestTask.toString() }

    /** 任务是否处于活动状态 */
    val isActive: Boolean
        get() {
            val job = _job ?: return false
            return job.isActive
        }

    /**
     * 提交任务
     */
    fun submit() {
        submitInternal()
    }

    @Synchronized
    private fun submitInternal() {
        if (_job != null) return

        _job = GlobalScope.launch(Dispatchers.Main) {
            try {
                // 通知开始
                withContext(Dispatchers.Main) {
                    notifyStart()
                }

                // 发起请求
                val response: IResponse = withContext(Dispatchers.IO) {
                    try {
                        HttpLog.i("$_logPrefix execute ${Thread.currentThread().name}")
                        _request.execute()
                    } catch (e: Exception) {
                        notifyError(e)
                        null
                    }
                } ?: return@launch

                // 通知请求结果
                val notifyResponse: Boolean = withContext(Dispatchers.IO) {
                    try {
                        HttpLog.i("$_logPrefix notifyResponse ${Thread.currentThread().name}")
                        _requestCallback.notifyResponse(response)
                        true
                    } catch (e: Exception) {
                        notifyError(e)
                        false
                    }
                }

                if (notifyResponse) {
                    // 通知成功
                    notifySuccess()
                }

            } catch (e: CancellationException) {
                notifyCancel()
            } finally {
                notifyFinish()
            }
        }
    }

    /**
     * 取消任务
     * @return true-任务被取消
     */
    @Synchronized
    fun cancel(): Boolean {
        val job = _job ?: return false
        if (!job.isActive) return false

        val isResultNotified = _isResultNotified
        HttpLog.e("$_logPrefix cancel start isStartNotified:${_isStartNotified} isResultNotified:${isResultNotified}")
        if (isResultNotified) return false

        job.cancel()
        val isActive = job.isActive
        val isCancelled = !isActive

        if (isCancelled && !_isStartNotified) {
            HttpUtils.runOnUiThread {
                notifyCancel()
                notifyFinish()
            }
        }

        HttpLog.e("$_logPrefix cancel finish isActive:${isActive}")
        return isCancelled
    }

    private fun notifyStart() {
        HttpLog.i("$_logPrefix notifyStart ${Thread.currentThread().name}")
        _isStartNotified = true
        _requestCallback.notifyStart()
    }

    private suspend fun notifyError(e: Exception) {
        require(e !is CancellationException)
        withContext(Dispatchers.Main) {
            HttpLog.i("$_logPrefix notifyError:$e ${Thread.currentThread().name}")
            _isResultNotified = true
            _requestCallback.notifyError(HttpException.wrap(e))
        }
    }

    private fun notifySuccess() {
        HttpLog.i("$_logPrefix notifySuccess ${Thread.currentThread().name}")
        _isResultNotified = true
        _requestCallback.notifySuccess()
    }

    private fun notifyCancel() {
        HttpLog.i("$_logPrefix notifyCancel ${Thread.currentThread().name}")
        _isResultNotified = true
        _requestCallback.notifyCancel()
    }

    private fun notifyFinish() {
        HttpLog.i("$_logPrefix notifyFinish ${Thread.currentThread().name}")
        _requestCallback.notifyFinish()
        this@RequestTask.onFinish()
    }

    abstract fun onFinish()
}